x/protocolpool/keeper/keeper.go

```diff
@@ -90,6 +107,112 @@ func (k Keeper) GetCommunityPool(ctx context.Context) (sdk.Coins, error) {
+	totalRecipientBal := recipientAmount + fundsAllocated
+	// check if the recipient account balance exceeds maxDistributedCapital after distribution
+	var withdrawnAmount sdk.Coin
+	if totalRecipientBal < cf.MaxDistributedCapital {
+		withdrawnAmount = sdk.NewCoin(denom, math.NewIntFromUint64(fundsAllocated))
+		// Distribute funds to the recipient from pool module account
+		err := k.DistributeFromCommunityPool(ctx, sdk.NewCoins(withdrawnAmount), recipient)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+
+		// decrement fundsAllocated from toDistribute
+		toDistribute -= fundsAllocated
+	}
+
+	return withdrawnAmount, nil
+}
+
+func (k Keeper) iterateAndUpdateFundsDistribution(ctx context.Context) (string, error) {
+	var totalPercentageToBeDistributed uint64
+	err := k.RecipientFundPercentage.Walk(ctx, nil, func(key sdk.AccAddress, value uint64) (stop bool, err error) {
+		totalPercentageToBeDistributed += value
+		return false, nil
+	})
+	if err != nil {
+		return "", err
+	}
+	if totalPercentageToBeDistributed > 100 {
+		return "", fmt.Errorf("total funds percentage cannot exceed 100")
+	}
+
+	poolMAcc := k.authKeeper.GetModuleAccount(ctx, types.ModuleName)
+
+	distrMAcc := k.authKeeper.GetModuleAccount(ctx, distrtypes.ModuleName)
+	distrBal := k.bankKeeper.GetAllBalances(ctx, distrMAcc.GetAddress())
+	distrDecAmount := sdk.NewDecCoinsFromCoins(distrBal...)
+
+	var denom string
+	err = k.RecipientFundPercentage.Walk(ctx, nil, func(key sdk.AccAddress, value uint64) (stop bool, err error) {
+		// Calculate the funds to be distributed based on the percentage
+		distributionAmount := distrDecAmount.MulDec(math.LegacyNewDecWithPrec(int64(value), 2))
+		denom = distributionAmount.GetDenomByIndex(0)
+		distrAmount := distributionAmount.AmountOf(denom)
+		distrCoins := distrAmount.TruncateInt().Uint64()
+
+		// Send distribution funds to pool module account [i.e., poolMAcc = MAcc + toDistribute]
+		err = k.bankKeeper.SendCoinsFromModuleToModule(ctx, distrMAcc.GetName(), poolMAcc.GetName(), distrBal)
+		if err != nil {
+			return false, err
+		}
+		// Add all the coins to be distributed to toDistribute
+		toDistribute += distrCoins
+
+		// Set funds to be claimed
+		toClaim, err := k.RecipientFundDistribution.Get(ctx, key)
+		if err != nil {
+			return false, err
+		}
+		amount := toClaim + distrCoins
+		err = k.RecipientFundDistribution.Set(ctx, key, amount)
+		if err != nil {
+			return false, err
+		}
+		return false, nil
+	})
+	return denom, err
+}
+
+func (k Keeper) getDistributedFunds(ctx context.Context, recipient sdk.AccAddress) (amount uint64, err error) {
+	amount, err = k.RecipientFundDistribution.Get(ctx, recipient)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return 0, fmt.Errorf("no recipient fund found for recipient: %s", recipient.String())
+		}
+		return 0, err
+	}
+
+	// set claimable to zero
+	err = k.RecipientFundDistribution.Set(ctx, recipient, 0)
```

Comment:
If we're writing to/modifying state, the method should not be called `get*` as that implies it's idempotent.